flocker/node/agents/blockdevice.py

```diff
@@ -617,4 +871,32 @@ def calculate_necessary_state_changes(self, local_state,
             for manifestation
             in manifestations_to_create
         )
-        return InParallel(changes=creates)
+
+        deletes = self._calculate_deletes(configured_manifestations)
+
+        return InParallel(changes=creates + deletes)
+
+    def _calculate_deletes(self, configured_manifestations):
+        """
+        :param dict configured_manifestations: The manifestations configured
+            for this node (like ``Node.manifestations``).
+
+        :return: A ``list`` of ``DestroyBlockDeviceDataset`` instances for each
+            volume that needs to be destroyed based on the given
+            configuration and the actual state of volumes (ie which exist
+            and which don't).
+        """
+        volumes = self.block_device_api.list_volumes()
```

Comment:
This seems bad... `calculate_etc` doesn't return Deferred because the presumption is that it's an in-memory operation, purely data structure transformation. I.e. it's presumed that local state, cluster state and cluster configuration are all you need. `DestroyBlockDeviceDataset` can do the dataset_id -> device mapping itself; it's even more inefficient, but we'd want to optimize this later anyway and we may as well go for correctness.